credentials/local/local.go

```diff
@@ -0,0 +1,104 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package local includes implementation of local credentials.
+// The local credential should be used in either a local TCP or UDS connection.
+// This package is experimental.
+package local
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"strings"
+
+	"google.golang.org/grpc/credentials"
+)
+
+// Info contains the auth information for a local connection.
+// It implements the AuthInfo interface.
+type Info struct {
+	credentials.CommonAuthInfo
+}
+
+// AuthType returns the type of Info as a string.
+func (t Info) AuthType() string {
+	return "local"
+}
+
+// localTC is the credentials required to eatablish a local connection.
+type localTC struct {
+	info *credentials.ProtocolInfo
+}
+
+func (c *localTC) Info() credentials.ProtocolInfo {
+	return *c.info
+}
+
+// getSecurityLevel returns the security level for a local connection.
+// It returns an error if a connection is not local.
+func getSecurityLevel(network, addr string) (credentials.SecurityLevel, error) {
+	switch {
+	// Local TCP connection
+	case strings.HasPrefix(addr, "127."), strings.HasPrefix(addr, "localhost:"), strings.HasPrefix(addr, "[::1]:"):
```

Comment:
Is addr the address as passed to `net.Dial`?

If so, the following will not work:

```
/etc/hosts:
127.0.0.1 workstation

foo.go:
grpc.Dial("workstation", local.NewCredentials())
```

Maybe this is fine, but I wanted to point that out.  If there's a way to retrieve the actual IP address, that would be better.

Reply:
No, it's not the address passed to `net.dial` but the one derived from `RemoteAddr().String()` of an established `net.Conn`, which I believe is the actual IP address.

Reply:
If that is the case, then is "localhost:" needed?